From e5695e846f2fb6de6b408526a6af7ca311daff3b Mon Sep 17 00:00:00 2001
From: Roman Rakus <rrakus@redhat.com>
Date: Tue, 14 Feb 2012 16:54:53 +0100
Subject: [PATCH 7/9] Handle Extended crashkernel syntax

see kernel/Documentation/kdump/kdump.txt
Correctly parse Extended crashkernel syntax - if the system has memory
in some range, use value from that range. Afterward save value
in basic crashkernel syntax

Signed-off-by: Roman Rakus <rrakus@redhat.com>
---
 src/system-config-kdump.py |   20 +++++++++++++++++++-
 1 files changed, 19 insertions(+), 1 deletions(-)

diff --git a/src/system-config-kdump.py b/src/system-config-kdump.py
index c985822..cea8cd8 100755
--- a/src/system-config-kdump.py
+++ b/src/system-config-kdump.py
@@ -599,7 +599,25 @@ class MainWindow:
                 crash_string = filter(lambda t: t.startswith("crashkernel="), 
                                      cmdline.split())[0].split("=")[1]
                 tokens = crash_string.split("@")
-                kdump_mem = int(tokens[0][:-1])
+                # Handle also Extended crashkernel syntax
+                def ws2mb(s):
+                    mult = 1
+                    if s[-1:] == "G":
+                        mult = 1024
+                    return int(s[:-1]) * mult
+
+                for rng in tokens[0].split(","):
+                    va = rng.split[":"]
+                    if len(va) == 1: # old syntax
+                        kdump_mem = ws2mb(va[0])
+                        break
+                    else:
+                        rmin, rmax = va[0].split(":")
+                        rmin = ws2mb(rmin)
+                        rmax = ws2mb(rmax)
+                        if total_mem >= rmin and total_mem < rmax:
+                            kdump_mem = ws2mb(va[1])
+                            break
                 if len(tokens) == 2:
                     kdump_offset = int(tokens[1][:-1])
 
-- 
1.7.7.6

